#37 - Display Name Validation

wey-chiang opened the pull request:
Added imports And, Regex from schemas. Added character validation for DisplayName, GlobalID, PolicyID, and RuleID

### Background

Panther does not accept certain special characters in display names, such as:

'
"
&
,
and perhaps others. The panther_analysis_tool should warn against these characters when they are present in a displayname.

### Changes

All changes, with the exception of related testing, that have been made are specific to the `panther_analysis_tool/schemas.py` file.
Created a new global variable `NAME_ID_VALIDATION_REGEX` which is meant to check against strings to ensure the string only contains alphanumeric characters along with dashes, underscores, periods, and spaces.
For GLOBAL_SCHEMA, POLICY_SCHEMA, and RULE_SCHEMA, the schema definition for the fields DisplayName, GlobalID, PolicyID, and RuleID have been updated to verify that the field value is both: of type str and matching NAME_ID_VALIDATION_REGEX

### Testing

Created a copy of tests/fixtures/valid_analysis/policies/example_policy.yml and the respective .py file
Edited the copy to include invalid characters
![image](https://user-images.githubusercontent.com/71197790/93529304-ac5cb480-f909-11ea-92d0-08b9cd9bd7a8.png)

austinbyers commented on panther_analysis_tool/schemas.py:
```diff
@@ -14,7 +14,9 @@
-from schema import Optional, Or, Schema
+from schema import And, Optional, Or, Regex, Schema
+
+NAME_ID_VALIDATION_REGEX = Regex(r"^[A-Za-z0-9\-\_\.\ \(\)]+$")
```
Nit: inside a character group you typically don't need escaping. This should be equivalent:

```suggestion
NAME_ID_VALIDATION_REGEX = Regex(r"^[A-Za-z0-9_. ()-]+$")
```

A little easier to read without all the escapes. (Note the `-` is now at the end so it will be interpreted literally.)

austinbyers reviewed (commented):
Deferring final approval to @nhakmiller , but looks good to me!

For context, we ban those characters to prevent HTML injection attacks

nhakmiller reviewed (commented):
Looks good! Love this clean implementation. Could you add a unit test (in `tests/unit/panther_analysis_tool`) to check for this case? You can add new test cases in `test/fixtures` as needed.

wey-chiang commented:
> Could you add a unit test (in `tests/unit/panther_analysis_tool`) to check for this case? 

Sure! I've added `example_policy_invalid_characters.py` and `example_policy_invalid_characters.yml` to the fixtures dir

wey-chiang commented:
> Looks good! Love this clean implementation. Could you add a unit test (in `tests/unit/panther_analysis_tool`) to check for this case? You can add new test cases in `test/fixtures` as needed.

Sure! I've added `example_policy_invalid_characters.py` and `example_policy_invalid_characters.yml` to the fixtures dir and updated `unit/panther_analysis_tool/test_main.py` accordingly.

Note that for `test_invalid_rule_definition` the `assert_equal(len(invalid_specs), 1)` statement had to be incremented e.g. `assert_equal(len(invalid_specs), 2)` with the new addition of the `example_policy_invalid_characters`

nhakmiller reviewed (approved):
LGTM
